src/protocol/sort/shuffle.rs

```diff
@@ -0,0 +1,100 @@
+use crate::helpers::prss::PrssSpace;
+use rand::rngs::mock::StepRng;
+use rand::seq::SliceRandom;
+
+#[allow(dead_code)]
+pub struct Shuffle {}
+
+impl Shuffle {
+    /// This implements Fisher Yates shuffle described here <https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle>
+    #[allow(clippy::cast_possible_truncation, dead_code)]
+    fn generate_random_permutation(
+        batchsize: usize,
+        index: u64,
+        left_permute: bool,
+        prss: &PrssSpace,
+    ) -> Vec<usize> {
+        let mut permutation: Vec<usize> = (0..batchsize).collect();
+        let rand = prss.generate_values(index.into());
+        let mut rng = if left_permute {
+            StepRng::new(rand.0 as u64, 1)
+        } else {
+            StepRng::new(rand.1 as u64, 1)
+        };
+        permutation.shuffle(&mut rng);
```

Comment:
The test seems to hang for 5-10secs while executing this function of  permutation.shuffle! Not sure whats causing this. Do you think using a better random generator might help here? If so, which one?
@akoshelev @thurstonsand @taikiy

Reply:
That's concerning. Randomly shuffling 10000 elements shouldn't take 5-10s.

Reply:
[`StepRng`](https://docs.rs/rand/latest/rand/rngs/mock/struct.StepRng.html) is not intended to be used in prod code, it's main purpose is to provide predictable rng for testing purpose.

the issue you're observing is caused by uniform sampler - because `StepRng` returns back numbers that are not uniformly distributed across the range, it causes more rejections by the uniform sampler. [this loop](https://docs.rs/rand/latest/src/rand/distributions/uniform.rs.html#547-553) is the hot spot.

I am curious whether we can let each helper to generate random permutation without any coordination with other helpers? In this case using a crypto rng here should suffice

Reply:
We need the seed to be the same for the peers so that they generate the exact same permutation. This would need prss

Reply:
can you generate a seed using prss and then initialize crypto rng with that seed?

Reply:
I think that should work.

On line 21 would it just be:
    CryptoRng::new(rand.0 as u64, 1)

And on line 23 just:
    CryptoRng::new(rand.1 as u64, 1)

Reply:
yep just want remark that we should use all 16 bytes of randomness provided by PRSS to initialize it (we talked with @richajaindce about it - PRSS uses AES256 internally so ideally we should be using 32 bytes, but that would require adding more methods to PRSS itself) - when @martinthomson is back, we can get some clarity on whether this approach is viable or not (using CryptoRng seeded by PRSS)

Reply:
So it seems like the PRNG is drawn on pretty heavily here, which means that any quality issues will turn into performance issues.  Fisher-Yates draws progressively smaller random values, which means that you draw at least n-1 times, starting at a draw of n, followed by n-1, n-2, etc...  The rand implementation uses rejection sampling, meaning that when you get to 2^x-1 you have a 50% chance of a rejection and another draw.

With `StepRng`, you are stepping one at a time through the entire output space, starting from the initial value.  Thus, once you exceed the current value, you need to go right around and then back to the start.  Little surprise then that this is slow as it takes almost 2^64 operations for each draw.

The PRSS implementation uses modulo.  This might be faster for this use case (I don't know).  It will at least be more deterministic, though it might not be constant time.

Seeding AES-256 requires 32 bytes, but you should look to do that from the shared entropy (i.e., spawn a new space) rather than seeding from its own output.

Reply:
I was thinking of using a Chacha8Rng which needs a seed of size [u8;32]. However, PRSS gives a seed size u128 (equivalent to [u8;16]. In this case, 
1. we can copy [u8;16] twice or 
2. generate two randoms from prss (second random generated from the returned index hence not enough entropy) and append them
I am implementing first approach for now. What do you think?

Reply:
Chach8Rng expects a [u8;32] seed. This is also the recommended size from the docs. Since PRSS is currently giving [u8;16], we can do following
1. Use u8;16 twice
2. have two PrssSpace to give us two random seeds.
For 2nd, ideally we should have a Space which returns a [u8;32] (which internally instantiates two spaces). Which approach are we comfortable taking? @martinthomson @akoshelev

Reply:
Also implemented 1st approach since that looks the cleanest for now. Based on discussion with @akoshelev there is an option of creating fake steps for protocols which need to use prss to generate seeds but will make the code un-necessarily complicated